### ISSUE TITLE
Compatibility of `DeviceArray` with `PIL.Image.fromarray()` to mirror numpy behavior

### ISSUE THREAD

**Issue by josephrocca** [google/jax#7932]
This is just a "quality of life" feature request, but twice now I've had to import the real `numpy` to do this sort of thing:
```python
PIL.Image.fromarray(np.array(jnp_array)).show()
```
It would be nice to be able to avoid importing the real numpy and just write:
```python
PIL.Image.fromarray(jnp_array).show()
```
Apologies if this is actually trivially achievable without importing the real numpy - I'm new to Python. Seems possible since no one has raised this issue yet. But then again, it'd be nice to have compatibility if it's possible, since this `Image.fromarray(numpy_array).show()` pattern seems common?

Thanks!

**Comment by apaszke** [comment]
I don't think that we can do anything about it on the JAX side. JAX arrays can already be converted to NumPy arrays because they implements the `__array__` magic method ([described here](https://numpy.org/devdocs/user/basics.dispatch.html)). The real fix would be to make PIL call `np.asarray` on the argument, as many Python libraries do.

**Comment by jakevdp** [comment]
We may be able to enable this from the JAX side, as it looks like PIL accepts objects that implement the Python side of the [array interface](https://numpy.org/doc/stable/reference/arrays.interface.html): https://github.com/python-pillow/Pillow/blob/869b786db158c869d0b771d4f546b6cf1d24e009/src/PIL/Image.py#L2825

That said, I'm not sure it's appropriate for JAX to implement that, because it brings the expectation of supporting direct CPU buffer access to the underlying data, which is not possible for JAX arrays in general.

**Comment by josephrocca** [comment]
@apaszke Ah, I got `'DeviceArray' object has no attribute '__array_interface__'` so I assumed this was something on JAX's end? 
```
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-13-b288747bbcb7> in <module>
     62 
     63 
---> 64 Image.fromarray(np.array(img)).show()

~/.local/lib/python3.6/site-packages/PIL/Image.py in fromarray(obj, mode)
   2816     .. versionadded:: 1.1.6
   2817     """
-> 2818     arr = obj.__array_interface__
   2819     shape = arr["shape"]
   2820     ndim = len(shape)

AttributeError: 'DeviceArray' object has no attribute '__array_interface__'
```
Please feel free to close if there's nothing that can be done here. Thanks!

**Comment by hawkinsp** [comment]
JAX doesn't implement `__array_interface__`, but it does implement the PEP 3118 buffer protocol which is extremely similar. We could support both, most likely.

**Comment by hawkinsp** [comment]
Although perhaps PIL would do better to support the PEP 3118 buffer protocol, if it doesn't already...

**Comment by jakevdp** [comment]
Maybe @shoyer has thoughts about whether we should implement both of these interfaces, or whether this is worth a bug against PIL.

**Comment by shoyer** [comment]
I see no reason why PIL should not handle `__array__`. This would also give them out of the box support for other array-convertible types in Python like those from Xarray and Dask, which seems nice.

On the other hand, I guess JAX could also easily implement `__array_interface__`, likely by converting into a NumPy array and then calling `ndarray. __array_interface__()`.
